### DISCUSSION TITLE
Add the keyword to load the type the interface is implemented

### DISCUSSION THREAD

**Discussion by NagayamaToshiaki** [dotnet/csharplang#6831] [Language Ideas] [answered]
In an interface, when you need to declare a method that calls the type it is implemented, you need to use generic.

```
interface IFoo<TFoo>
{
    TFoo DoSomething();
}
```

I guess this is largely due to legacy non-generic interfaces such as `IComparable`.

This syntax has a couple of problems:

- You can't ensure `TFoo` is the type that implements IFoo. You can write `where TFoo: IFoo<TFoo>` to tell what you meant, but it doesn't tell IDE to stop a different type is loaded.
- It's verbose. You need to type the type name again. Also, It's quite often no use you set different type other than its own.

So the need for a simple declaration is high. I'd like `IFoo` can be written like this:

```
interface IFoo
{
    ImplimentationType DoSomething();
}
```

I know this is a very distractive change, but rarely the type is named `ImplimentationType`.

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/blob/main/proposals/self-constraint.md

see also [exsitential types](https://github.com/dotnet/csharplang/issues/5556)
